lib/ddtrace/contrib/grpc/datadog_client_interceptor.rb

```diff
@@ -0,0 +1,53 @@
+require 'grpc'
+require 'ddtrace/propagation/grpc_propagator'
+
+module Datadog
+  module Contrib
+    module GRPC
+      # The DatadogClientIntercptor implements the basic interface
+      # for gRPC middleware. When you declare Datadog usage of the gRPC
+      # module, the interceptor must be explicitly declared for usage.
+      class DatadogClientInterceptor < ::GRPC::ClientInterceptor
+        def request_response(request: nil, call: nil, method: nil, metadata: nil)
+          trace_client('client', metadata) { yield }
+        end
+
+        def client_streamer(requests: nil, call: nil, method: nil, metadata: nil)
+          trace_client('client', metadata) { yield }
+        end
+
+        def server_streamer(request: nil, call: nil, method: nil, metadata: nil)
+          trace_client('client', metadata) { yield }
+        end
+
+        def bidi_streamer(requests: nil, call: nil, method: nil, metadata: nil)
+          trace_client('client', metadata) { yield }
+        end
+
+        private
+
+        def ddtracer
+          @ddtracer ||= @options.fetch(:ddtracer, Datadog.tracer)
+        end
+
+        def trace_client(name, metadata = {})
+          metadata ||= {}
+          tracer_options = {
+            service: Datadog.configuration[:grpc][:service_name],
+            span_type: 'grpc'
+          }
+          span = ddtracer.trace(name, tracer_options)
+          Datadog::GRPCPropagator.inject!(span.context, metadata)
+
+          yield
```

Comment:
Since an interceptor is like middleware, does this `yield` equivocate to the actual gRPC call? In which case, could we wrap the `trace` method around it like:

```
ddtracer.trace(name, tracer_options) do |span|
  yield
end
```

The benefit you'd get here is automatic `span.set_error` and `span.finish` functions rendering the `rescue` and `ensure` blocks as moot.